### PR TITLE
[code-quality] tests/vm_test.go: Replace `Update()` with `Patch()`

### PR DESCRIPTION
**Pull request by orenc1**
Replacing `Update()` directives in tests/vm_test.go with `Patch()` in order to avoid conflicts on objects and reduce flakiness.

### What this PR does
Before this PR:
`.Update()` is being used to update VM objects.
After this PR:
`.Patch()` is being used to update VM objects according to the test cases.

Fixes #
Jira-issue: https://issues.redhat.com/browse/CNV-44590

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
tests/vm_tests.go: replace Update() with Patch()
```

### CODE REVIEW

File: tests/vm_test.go [hunk 1]
```diff
@@ -289,15 +289,18 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
-				vm.Spec.Template.ObjectMeta.Annotations = annotations
-				vm, err = virtClient.VirtualMachine(meta.Namespace).Update(context.Background(), vm, metav1.UpdateOptions{})
-				return err
-			})
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			annotationsPatch, err := patch.New(
+				patch.WithAdd(
+					"/spec/template/spec/metadata/annotations",
+					annotations,
```
Comment (EdDev) [inline tests/vm_test.go:297]: nit: Can these be placed on the same line? It is a bit noisy to use newlines when not really needed.

File: tests/vm_test.go [hunk 2]
```diff
@@ -289,15 +289,18 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+				)).GeneratePayload()
 			Expect(err).ToNot(HaveOccurred())
 
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, annotationsPatch, metav1.PatchOptions{})
+
```
Comment (EdDev) [inline tests/vm_test.go:302]: nit: The `Expect` fits immediately after the operation without a newline.
The newline is after the `Expect` to separate it from the next operation of starting the VM.

File: tests/vm_test.go [hunk 3]
```diff
@@ -289,15 +289,18 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
 			vm := createVM(virtClient, libvmifact.NewCirros())
 
-			err = retryWithMetadataIfModified(vm.ObjectMeta, func(meta k8smetav1.ObjectMeta) error {
-				vm, err = virtClient.VirtualMachine(meta.Namespace).Get(context.Background(), meta.Name, k8smetav1.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				vm.Spec.Template.ObjectMeta.Annotations = annotations
-				vm, err = virtClient.VirtualMachine(meta.Namespace).Update(context.Background(), vm, metav1.UpdateOptions{})
-				return err
-			})
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, k8smetav1.GetOptions{})
```
Comment (EdDev) [inline tests/vm_test.go:292]: This is not needed.

File: tests/vm_test.go [hunk 4]
```diff
@@ -289,15 +289,18 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
...
+				patch.WithAdd(
+					"/spec/template/spec/metadata/annotations",
+					annotations,
+				)).GeneratePayload()
 			Expect(err).ToNot(HaveOccurred())
 
+			vm, err = virtClient.VirtualMachine(vm.Namespace).Patch(context.Background(), vm.Name, types.JSONPatchType, annotationsPatch, metav1.PatchOptions{})
+
+			Expect(err).ToNot(HaveOccurred())
 			vm = startVM(virtClient, vm)
```
Comment (EdDev) [inline tests/vm_test.go:304]: And now that I see this, it looks very similar to something we already handled in the past.
There is no need for and update or patch, the VM should be created with the annotation in the first place. There is no meaning of updating the VM after the creation and without it running yet... there is no VMI yet.

Please examine if this is the same for the other cases below.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: tests/vm_test.go (hunk 1 above)
Comment (orenc1) [reply tests/vm_test.go]: now it's on the same line when the total line length is reasonable.

**Review by EdDev — APPROVED** [review]
Thanks!

File: tests/vm_test.go [hunk 5]
```diff
@@ -347,9 +347,13 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 		It("should not update the vmi generation annotation when the template changes", func() {
 			vm := startVM(virtClient, createVM(virtClient, libvmifact.NewCirros()))
 
-			By("Updating the VM template spec")
-			vm.Spec.Template.ObjectMeta.Labels = map[string]string{"testkey": "testvalue"}
-			_, err := virtClient.VirtualMachine(vm.Namespace).Update(context.Background(), vm, metav1.UpdateOptions{})
+			By("Updating the VM template metadata")
+			labelsPatch, err := patch.New(
```
Comment (enp0s3) [inline tests/vm_test.go:351]: @orenc1 I think to be on the safe you should add the `patch.WithTest` in order verify existence of the original values in `/spec/template/metadata/labels`

Comment (EdDev) [reply tests/vm_test.go:351]: I do not think the test cares about someone else colliding. If it cares, then it should fail on something else (the generation state?).

but I do not have any strong opinions here.

Comment (orenc1) [reply tests/vm_test.go:351]: in this case, the VM is new and there shouldn't be any labels on it yet.
do you think it's neccesary to check it has no labels, just after VM creation?

**Comment by xpivarc** [comment]
>pull-kubevirt-check-tests-for-flakes — Job failed.  

/hold

**Comment by EdDev** [comment]
> > pull-kubevirt-check-tests-for-flakes — Job failed.
> 
> /hold Please have a look at the failure that seems related.

@xpivarc 
I tried to investigates and I do not see how it is related.
It fails on a different test for a different reason.

```yaml
{
                        "type": "Ready",
                        "status": "False",
                        "lastProbeTime": null,
                        "lastTransitionTime": "2024-07-31T14:47:13Z",
                        "reason": "ContainersNotReady",
                        "message": "containers with unready status: [guest-console-log]"
}
```

And it was suppose to be up.

Can you see a real relation to the change?
Can you help detect the reason the guest console log container failed to come up?

**Comment by orenc1** [comment]
@xpivarc , `pull-kubevirt-check-tests-for-flakes` has passed now.

**Comment by xpivarc** [comment]
/hold cancel
It seems the test is not adjusted. Sorry I am only looking at the lane, not code recently to prevent more flakiness. Looks good.

**Pull request merged**